atc/engine/check_delegate.go

```diff
@@ -106,22 +106,19 @@ func (d *checkDelegate) WaitToRun(ctx context.Context, scope db.ResourceConfigSc
 
 	var lock lock.Lock = lock.NoopLock{}
 	if d.plan.IsPeriodic() {
-		for {
-			var acquired bool
-			lock, acquired, err = scope.AcquireResourceCheckingLock(logger)
-			if err != nil {
-				return nil, false, fmt.Errorf("acquire lock: %w", err)
-			}
-
-			if acquired {
-				break
-			}
-
-			select {
-			case <-ctx.Done():
-				return nil, false, ctx.Err()
-			case <-d.clock.After(time.Second):
-			}
+		scopeLock, acquired, err := scope.AcquireResourceCheckingLock(logger)
+		if err != nil {
+			return nil, false, fmt.Errorf("acquire lock: %w", err)
+		}
+
+		// For Lidar checks or manually triggered checks, if not acquire lock,
+		// meaning there is a check on the resource is running, so that instead
+		// of waiting for current check to finish, we can just abort current
+		// check.
```

Comment:
Yes. I think store check result in `state` only useful for normal build as `state` is per build. Please correct me if I misunderstood.

Reply:
I think we may need to store the result always, but I'm not entirely sure. Maybe @clarafu has a better idea?

e.g. the "image check" must always produce a result
https://github.com/concourse/concourse/blob/3365f6622b7ae476e4b9daf92bf66afc625fdd4b/atc/engine/build_step_delegate.go#L287-L289

Reply:
Chatted with @clarafu about this some more, and we're pretty sure this'll cause issues, particularly with global resources disabled. Since the ResourceCheckingLock is on the resource config ID, but versions are per scope, we may end up not running a check for a resource for several check intervals (since the resource config ID may be locked by another resource). The above reason also applies (we always expect a check to store a version).

Is there a particular reason for this short-circuit besides just reducing the number of active Goroutines? I think it's important to keep the old behaviour here

Reply:
I think over here. My original thought was to quickly terminate a in-memory check build if it cannot acquire resource lock. But with `inMemoryCheckBuild.AcquireTrackingLock()`, it's ok to wait for lock here, because no duplicate checks on the same resource will run. I can revert this change.